repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomEntryStreamBuilder.scala

```diff
@@ -0,0 +1,105 @@
+
+import org.apache.abdera.Abdera
+import org.apache.abdera.model.{Entry, Feed}
+import org.openrepose.commons.utils.http.CommonHttpHeader
+import org.openrepose.commons.utils.logging.{TracingHeaderHelper, TracingKey}
+import org.openrepose.nodeservice.atomfeed.AuthenticatedRequestFactory
+import org.slf4j.MDC
+
+import scala.collection.JavaConversions._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.language.postfixOps
+
+/**
+  * A container object for the build function.
+  */
+object AtomEntryStreamBuilder {
+
+  private val parser = Abdera.getInstance().getParser
+
+  /**
+    * Calling this function will retrieve the first page of the Atom feed located at the provided baseFeedUrl.
+    * Subsequent pages will be retrieved lazily. To keep a reference to a Stream produced by this function without
+    * making a request to the Atom feed, assign the result of this function to a lazy variable like so:
+    * lazy val myFeed = build("http://my.feed.url", MyAuthenticatedRequestFactory)
+    *
+    * @param baseFeedUrl a static locator pointing to the "head" of an Atom feed (e.g., the subscription document)
+    * @param authenticator a URL processor which authenticates connections
+    * @param authenticationTimeLimit a maximum [[Duration]] to wait on authentication before considering authentication
+    *                                a failure
+    * @return a Scala [[scala.collection.immutable.Stream]] which will yield Atom entries in an Atom feed located at the
+    *         provided baseFeedUrl. The entries will be yielded in the order in which they are read from the feed XML,
+    *         from top to bottom.
+    * @throws AuthenticationException when the authenticator fails to authenticate a request
+    */
+  def build(baseFeedUrl: URL,
+            reposeVersion: String,
+            authenticator: Option[AuthenticatedRequestFactory] = None,
+            authenticationTimeLimit: Duration = 1 second): Stream[Entry] = {
+    val requestId = java.util.UUID.randomUUID().toString
+    MDC.put(TracingKey.TRACING_KEY, requestId)
+    buildR(baseFeedUrl, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+  }
+
+  private def buildR(baseFeedUrl: URL,
+                     reposeVersion: String,
+                     authenticator: Option[AuthenticatedRequestFactory] = None,
+                     authenticationTimeLimit: Duration = 1 second,
+                     requestId: String): Stream[Entry] = {
+    val baseFeedConnection = baseFeedUrl.openConnection()
+
+    val authenticatedConnection = authenticator match {
+      case Some(arf) =>
+        val connectionFuture = Future(arf.authenticateRequest(baseFeedConnection, requestId, reposeVersion))
+        Option(Await.result(connectionFuture, authenticationTimeLimit))
+      case None =>
+        Some(baseFeedConnection)
+    }
+
+    authenticatedConnection match {
```

Comment:
There may have been a better reason, but writing this was so long ago I can only guess at this point.

My guess is that the Abdera parser takes an InputStream, Reader, or a couple of other things. I know we can get that out of our service clients, but I might not have made the connection. Additionally, the HttpConnectionPoolService returns an Apache HttpClient while the AkkaServiceClient takes a URI and headers as explicit parameters. In the former case, that requires the AtomFeedService to either force an Apache dependency on anyone writing an AuthenticatedRequestFactory or to translate between a URLConnection and the Apache class, while in the latter case, I suspect we don't have the ability to manipulate the request arbitrarily since the AkkaServiceClient uses Apache's client underneath while only providing access to two parameters for one method call.

URL connection manipulation provides the most freedom, but we lose some of the utility provided by our other service clients. I /think/ I can get the Akka service to work, but doing so might compel me to change the AuthenticatedRequestFactory interface. I'd like to talk about this more in person so I can scattershot some thoughts, ideas, and concerns.